hibernate-core/src/main/java/org/hibernate/type/format/StringJsonDocumentReader.java

```diff
@@ -0,0 +1,505 @@
+			if (c == escape) {
+				escapeIsOn = true;
+			}
+			else {
+				if ( c == character ) {
+					if (escapeIsOn) {
+						escapeIsOn = false;
+					}
+					else {
+						found = this.json.position() - 1;
+						break;
+					}
+				}
+			}
+		}
+		this.json.reset();
+		return found;
+	}
+
+	/**
+	 * Consume a non-quotted value
+	 * @return the length of this value. can be 0, -1 in case of error
+	 */
+	private int consumeNonStringValue() {
+		int newViewLimit = 0;
+		boolean allGood = false;
+		for (int i =  this.json.position(); i < this.json.limit(); i++ ) {
+			char c = this.json.get(i);
+			if ((StringJsonDocumentMarker.markerOf( c ) != StringJsonDocumentMarker.OTHER) ||
+				Character.isWhitespace( c )) {
+				// hit a JSON marker or a space.
+				allGood = true;
+				// give that character back to the buffer
+				newViewLimit = i;
+				break;
+			}
+		}
+
+		if (allGood) {
+			this.jsonValueWindow.limit(newViewLimit);
+			this.jsonValueWindow.position( this.json.position() );
+			this.json.position(newViewLimit);
+		}
+		return allGood?(this.jsonValueWindow.remaining()):-1;
+	}
+	private void consumeQuottedString() {
+
+		// be sure we are at a meaningful place
+		// key name are unquoted
+		moveTo( StringJsonDocumentMarker.QUOTE.getMarkerCharacter() );
+
+		// skip the quote we are positioned on.
+		this.json.get();
+
+		//locate ending quote
+		int endingQuote = locateCharacter( StringJsonDocumentMarker.QUOTE.getMarkerCharacter(), '\\');
+		if (endingQuote == -1) {
+			throw new IllegalStateException("Can't find ending quote of key name");
+		}
+
+		this.jsonValueWindow.limit( endingQuote );
+		this.jsonValueWindow.position(this.json.position());
+		this.json.position( endingQuote + 1);
+
+	}
+
+	private void ensureValueState() throws IllegalStateException {
+		if ((this.processingStates.getCurrent() !=  PROCESSING_STATE.OBJECT ) &&
+			this.processingStates.getCurrent() !=  PROCESSING_STATE.ARRAY)  {
+			throw new IllegalStateException( "unexpected processing state: " + this.processingStates.getCurrent() );
+		}
+	}
+	private void ensureAvailableValue() throws IllegalStateException {
+		if (this.jsonValueWindow.limit() == 0 ) {
+			throw new IllegalStateException( "No available value");
+		}
+	}
+
+	@Override
+	public String getObjectKeyName() {
+		if (this.processingStates.getCurrent() !=  PROCESSING_STATE.OBJECT_KEY_NAME ) {
+			throw new IllegalStateException( "unexpected processing state: " + this.processingStates.getCurrent() );
+		}
+		ensureAvailableValue();
+		return this.jsonValueWindow.toString();
+	}
+	@Override
+	public String getStringValue() {
+		ensureValueState();
+		ensureAvailableValue();
+		if (hasEscape(this.jsonValueWindow)) {
+			return unescape(this.jsonValueWindow);
+		}
+		return this.jsonValueWindow.toString();
+	}
+
+
+	@Override
+	public BigDecimal getBigDecimalValue() {
+		ensureValueState();
+		ensureAvailableValue();
+		return BigDecimal.valueOf( Long.valueOf(this.jsonValueWindow.toString()) );
```

Comment:
This looks wrong to me. Why don't you use this instead?

```suggestion
		return new BigDecimal( this.jsonValueWindow.toString() );
```